### ISSUE TITLE
A bad designed Class can cause an Exception when using Stringify with JsonFormatter or JsonValueFormatter

### ISSUE THREAD

**Issue by rafaelsc** [serilog/serilog#1451]
**Description**
A bad designed Class can cause an Exception when logging with stringification operator. Using the JsonFormatter or a Formatter that uses the JsonValueFormatter internally like CompactJsonFormatter or RenderedCompactJsonFormatter.

**Reproduction**
Log using stringification operator with a instance of a class that can return null in ToString(). With configured to use JsonFormatter or CompactJsonFormatter or RenderedCompactJsonFormatter.

```csharp
	Log.Logger = new LoggerConfiguration()
		.WriteTo.Console(new Serilog.Formatting.Json.JsonFormatter()) //Or other JsonFormatter
		.CreateLogger();

        class ABadBehavior
        {
            public override string ToString() => null;
        }
        Log.Information("{$Data}", new ABadBehavior());
```

**Expected behavior**
That the Serilog Core doesn't throw an exception. And Log the result like was another null property.

**Relevant package, tooling and runtime versions**
All runtimes and using the lastest Serilog codebase.

**Additional context**
I find this issue when implementing the C#8 Nullable Reference Type in Serilog codebase. 
Related Issue #1426

**Comment by nblumhardt** [comment]
Thanks for the note!

It's important to note whether the exception is thrown as far as the caller of the `Log.*` method, or if it just reaches `SelfLog`. Logging the exception to `SelfLog` is currently the expected behavior; throwing it back to the log method caller would be a critical bug.

There's some precedent for catching/suppressing these; during capturing ("destructuring") property accessors that throw end up logged with the exception as the property value. Would be nice for the change here to be consistent.

We'd need to check whether the addition of a try/catch block impacts perf in the non-throwing case.

**Comment by rafaelsc** [comment]
@nblumhardt Sorry, I give a better look here. And this Bug doesn't happen with the sample code that I sent. And should not happen easily for any user.

This Bug was fixed in #1427 too. That bug fix solves this problem for destructuring a class, and  JsonFormatter/CompactJsonFormatter uses the same destructuring code that is now fixed. 
But a very similar bug can happen inside the JsonFormatter/JsonValueFormatter and can be triggered calling the Formatter directly. (See the Unit Test in #1452)

SelfLog has triggered correctly for `Log.*` methods. Ano no exception is thrown outside `Log.*` methods. 

In this 2 LINQPad 5 Files (zipped) is easy to see the Bug. [Serilog ToString Bug.zip](https://github.com/serilog/serilog/files/4846747/Serilog.ToString.Bug.zip)

In `Serilog ToString Bug - Last Stable` we see the SelfLog logging an error in the Last Stable version one time (in the destructuring phase) ![image](https://user-images.githubusercontent.com/502282/86031412-a500cb80-b9ea-11ea-9306-d95950665bdb.png)
And in `Serilog ToString Bug - Last PreRelease` we don't see the problem anymore because of the fix #1452. 
![image](https://user-images.githubusercontent.com/502282/86031498-be097c80-b9ea-11ea-9e14-a308499ab96b.png)


Because of that, I think this bug can be closed, but the PR #1451 is still valid and should be merged because the Internal exception can happen, but not in the same way that I initially though.
